kart/spatial_tree.py

```diff
@@ -0,0 +1,413 @@
+            if geom is None:
+                continue
+            try:
+                s2_cell_tokens = find_s2_cells(s2_coverer, geom, transforms)
+            except Exception as e:
+                L.warning(f"Couldn't locate S2 cells for {feature_oid}:\n{e}")
+                continue
+
+            params = (bytes.fromhex(feature_oid),)
+            row = dbcur.execute(
+                "SELECT rowid FROM blobs WHERE blob_id = ?;", params
+            ).fetchone()
+            if row:
+                rowid = row[0]
+            else:
+                dbcur.execute("INSERT INTO blobs (blob_id) VALUES (?);", params)
+                rowid = dbcur.lastrowid
+
+            if not s2_cell_tokens:
+                continue
+
+            params = [(rowid, token) for token in s2_cell_tokens]
+            dbcur.executemany(
+                "INSERT OR IGNORE INTO blob_cells (blob_rowid, cell_token) VALUES (?, ?);",
+                params,
+            )
+
+    t1 = time.monotonic()
+    click.echo(f"Indexed {i} features in {t1-t0:.1f}s")
+
+
+NO_GEOMETRY_COLUMN = object()
+
+
+def get_geometry(repo, feature_oid):
+    legend, fields = msg_unpack(repo[feature_oid])
+    col_id = get_geometry.legend_to_col_id.get(legend)
+    if col_id is None:
+        col_id = _find_geometry_column(fields)
+        get_geometry.legend_to_col_id[legend] = col_id
+    return fields[col_id] if col_id is not NO_GEOMETRY_COLUMN else None
+
+
+get_geometry.legend_to_col_id = {}
+
+
+def _find_geometry_column(fields):
+    for i, field in enumerate(fields):
+        if isinstance(field, Geometry):
+            return i
+    return NO_GEOMETRY_COLUMN
+
+
+def find_s2_cells(s2_coverer, geom, transforms):
+    is_point = geom.geometry_type == GeometryType.POINT
+
+    return (
+        _point_f2_cells(s2_coverer, geom, transforms)
+        if is_point
+        else _general_s2_cells(s2_coverer, geom, transforms)
+    )
+
+
+def _apply_transform(original, transform, overwrite_original=False):
+    if transform is None:
+        return original
+    result = original if overwrite_original else original.Clone()
+    result.Transform(transform)
+    return result
+
+
+def _point_f2_cells(s2_coverer, geom, transforms):
+    import pywraps2 as s2
+
+    g = gpkg_geom_to_ogr(geom)
+
+    one_transform = len(transforms) == 1
+    if not one_transform:
+        result = set()
+
+    for transform in transforms:
+        g_transformed = _apply_transform(g, transform, one_transform)
+        p = g_transformed.GetPoint()[:2]
+        s2_ll = s2.S2LatLng.FromDegrees(p[1], p[0]).Normalized()
+        s2_token = s2.S2CellId(s2_ll.ToPoint()).ToToken()
+        if one_transform:
+            return (s2_token,)
+        else:
+            result.add(s2_token)
+
+    return result
+
+
+def _general_s2_cells(s2_coverer, geom, transforms):
+    import pywraps2 as s2
+
+    e = geom_envelope(geom)
+    if e is None:
+        return ()  # Empty.
+
+    sw_src, ne_src = (e[0::2], e[1::2])
```

Comment:
split this into two lines, remove the stride slice. too hard to read

Reply:
Done